### PR TITLE
Implement send-side throttling

### LINKED ISSUE #1: Implement throttling
Currently, even with 1 connection, fastsync triggers OVH's DDoS protection system, which limits the through put significantly

We want to avoid these events if possible.

### PR DESCRIPTION
**Pull request by DavidVentura**
This change closes #1.

Implement sender-side throttling by keeping a sliding window of all transfers that happened within the last 1 second.

Considering only the last second will keep a tighter bound on the used bandwidth, but may be problematic for high latency connections. The sliding window could be configurable, if valuable.

Should I add some integration tests for this functionality?

Sending 100MB with `--max-bandwidth 20` to localhost

```
[104595456 / 104857600]  99.8% 19.82 MB/s, 0.0 minutes left
SEND-CHUNK ChunkHeader { file_id: FileId(0), offset: 104595456, len: 262144 } header_len=14
Accepted connection from 127.0.0.1:60082.
```

Sending 100MB with `--max-bandwidth 10` to localhost
```
[104595456 / 104857600]  99.8% 9.94 MB/s, 0.0 minutes left
SEND-CHUNK ChunkHeader { file_id: FileId(0), offset: 104595456, len: 262144 } header_len=14
Accepted connection from 127.0.0.1:48560.
```

without passing `--max-bandwidth`, still to localhost

```
[104595456 / 104857600]  99.8% 2127.13 MB/s, 0.0 minutes left
SEND-CHUNK ChunkHeader { file_id: FileId(0), offset: 104595456, len: 262144 } header_len=14
```

### CODE REVIEW

**Comment by ruuda** [comment]
Thanks for picking this up so quickly!

I scrolled through the diff quickly, but before we jump into the implementation, how should the rate limiting actually work?

<details>
<summary>Lots of open questions, but maybe now I am over-engineering something simple; unstructured brain dump</summary>

* As you noticed, we need to pick some time window to measure the rate over. If we make the window too large then there can still be spikes where we are too fast and DDoS protection might kick in. If we make the window too small then the transfer speed stops being well-defined.
* Currently we send relatively large chunks to minimize the overhead of the header and to leave it up to the kernel and network how to best split the data into packets. (The chunk size of 256 KiB is somewhat arbitrary, I never measured if there  is an optimum.) Should we limit by shrinking the chunks, by sending them less often, or both?
  * If we increase the interval between sends then probably the transfer speed measured over a small interval becomes spiky. I don’t know if that is a problem.
  * If we make the chunks too small then the header size overhead becomes significant.
  * A chunk is effectively a packet that we send over a stream which again splits it into packets. If the chunk is large then it doesn’t matter so much how it is split into packets, but if the chunk size approaches the packet size, is that a problem? (The pathological case is where every packet contains the tail of one chunk and then the head of the next chunk.)
 * A leaky bucket rate limiter is maybe an elegant way to go about it.
   * When it is time to send, see how many bytes are left in the bucket, then send that many.
   * It still works if we set some minimum chunk size. If there aren’t that many bytes in the bucket, we can compute exactly how long to sleep for before there are enough in the bucket.
   * But with multiple sending threads, does that create a thundering herd problem if all of them wait until exactly the moment when there is one chunk to send? If one thread keeps winning that race then effectively we are back to one connection again which defeats the purpose of this tool. Does it matter though? Not being able to saturate the link is not a problem if we don’t _want_ to saturate it. And even if on the single connection, speed drops temporarily due to a retransmit, it just means it gets to transfer more later.
   * It could be one bucket with some kind of fairness thing to ensure all connections get used, or we could split the rate limit for every new connection and give each thread its own rate limiter. The downside of the latter is that then the sends can all sync up and we get spiky transfers again. E.g. 4 threads can each send a chunk simultaneously and then sleep for a second until the rate limit resets, but we want one thread to send every 250ms.
 * We need to watch out on the sending side how we measure, because it sends immediately, but the receiver has to accept, so the sending thread blocks until the receiving end accepts. If we record the start time as when we start sending, then it looks like the initial transfer speed is very low.
</details>

> Should I add some integration tests for this functionality?

We don’t have any tests (or even CI) for this tool at the moment. To add some context, it was written in a rush and then we added some features, it was simple enough that if you get it wrong you would notice. But now that we are adding more features and it’s becoming more subtle, we should probably add some form of testing.

Maybe then we can extract the rate limiter into a pure part with a virtual clock that we can test, because if we have to run the real thing in a test that’s going to be a slow test and probably also flaky.

But let’s do that later, I think for now “we ran it and it works” is an acceptable way to test.

**Comment by DavidVentura** [comment]
As a comparison:

With the sliding window approach, we send at the maximum rate possible during the start of the measured interval, then send nothing until the next window starts; this technically could cause buffering if any part of the path between sender and receiver cannot handle the transfer burst (this becomes a natural ratelimiter anyway, but with TCP congestion control it could cause _lower_ bandwidth than our desired limit). In this case, windows need to be somewhat large (1~2s).

Using the leaky bucket approach, eventually we hit a point where there are no tokens left (otherwise the ratelimiter does not achieve anything), at which point we effectively send 1 data unit (chunk?) per 1 token refill interval. The windows are fairly small ( `block size / target bandwidth`)


I think the leaky bucket approach is better, though a bit harder to implement nicely, I'll give it a shot.

> We need to watch out on the sending side how we measure, because it sends immediately, but the receiver has to accept, so the sending thread blocks until the receiving end accepts. If we record the start time as when we start sending, then it looks like the initial transfer speed is very low.

We can send without blocking until we fill up the `net.core.wmem_max` sysctl, which is ~1 chunk in my system (213KiB)

**Comment by ruuda** [comment]
> though a bit harder to implement nicely

It should be very easy to implement, you only need an amount + last updated time. (It’s a bit harder if you want to do it with atomics, but let’s just put it in a mutex.)

**Comment by DavidVentura** [comment]
You were right, it wasn't that complicated :grin:

File: src/ratelimiter.rs
Comment (ruuda) [inline src/ratelimiter.rs:25]: I would make the interface something like this:

```rust
pub fn bytes_available(&self, now: Instant) -> f32;
pub fn consume_bytes(&mut self, now: Instant, num_bytes: f32);
pub fn time_until_bytes_available(&self, now: Instant, num_bytes: f32) -> Duration;
```

Passing in `now` externally is not strictly needed but it makes the entire thing pure which makes it easier to test because you can say do something like this:

```rust
let mut rl = RateLimiter::new(...);
let t0 = Instant::now();
t1 = t0 + rl.time_until_available_bytes(t0, 100.0);
assert!(rl.bytes_available(t1) >= 100.0);
```
without the test having to block for a long time.

File: src/ratelimiter.rs
Comment (ruuda) [inline src/ratelimiter.rs:19]: Typically the bucket starts out full, not empty. If it starts out empty, the very first send we try to do will block unnecessarily.

File: src/main.rs
Comment (DavidVentura) [reply src/main.rs:38]: MB/s = Megabytes = 10**6
MiB/s = Mibibytes = 2**20

File: src/ratelimiter.rs
Comment (DavidVentura) [reply src/ratelimiter.rs:19]: Initially I started with a full bucket, but doing that you can send 2x the bucket size over the first transfer window.

I'm not sure that it's an actual problem, but I thought that if we are limiting the max bandwidth, we shouldn't exceed the threshold (instead of starting at 2x and gradually trending downwards to the threshold)

What do you think?

File: src/main.rs
Comment (DavidVentura) [reply src/main.rs]: if we release the lock during sleep, another thread can show up and request the same amount of bytes.
I can implement the logic in a loop, but am not sure it's the best from a practical perspective: N threads will wake up in a loop and get blocked "forever" while only a subset of them make progress

what do you think?

File: src/ratelimiter.rs
Comment (DavidVentura) [reply src/ratelimiter.rs]: This is a lot nicer, though I did not end up using `bytes_available` within `fastsync`, only to test internal behavior

File: src/ratelimiter.rs
Comment (DavidVentura) [reply src/ratelimiter.rs]: just not making things more public than needed, but I've made it `pub`

File: src/main.rs
Comment (ruuda) [reply src/main.rs]: > N threads will wake up in a loop and get blocked "forever" while only a subset of them make progress

Yep, so now we get to the subtleties :grimacing: I have a feeling that we shouldn’t sleep while holding the lock because if a different thread needs to only read it to check something then it blocks. But at the same time I don’t think that is a problem right now and it would be the simplest thing to do.

If we don’t hold the lock while sleeping then we might get the thundering herd of threads all waking up but only one can consume the new available bytes. It’s only a few threads so I don’t expect overhead is a big problem, but it may lead to a fairness problem where one thread is winning all the time and the others are not sending. That is also not a problem in one sense, we’ll still send all the data, but it means we’re only utilizing one TCP connection. Which is maybe not a problem if we can saturate the target rate with one connection. But if we really want to do it properly, what we can do is add a `VecDeque` next to to the rate limiter, also protected by the lock, where we put the thread ids of waiting threads. Then if the limit is exhausted and you are not already in the queue, you put yourself at the end of the queue and go to sleep until you think the bucket is full enough already. When you wake up, if you are at the front of the queue then you consume from the limit, otherwise go back to sleep again.

It’s a lot more complex than sleeping with the lock, but sleeping with the lock does not guarantee fairness either because `std::sync::Mutex` does not guarantee fairness, so we can still have this situation where only one thread is sending all the time so we use only one TCP connection. Do we really need fairness though? Probably not but maybe we should measure this? Though that would be a lot of work and probably unfair is good enough …

File: src/ratelimiter.rs
Comment (ruuda) [reply src/ratelimiter.rs]: > but doing that you can send 2x the bucket size over the first transfer window.

I don’t think so, but it’s a bit subtle. I think pictures make it clearer. This is what an ideal transfer would look like:

send-ceil

Where on the y-axis is cumulative bytes sent (in arbitrary units, where 1 means 1 bucket, not necessarily 1 byte), and on the x-axis is time. It’s not so clear in the plot but the intervals should be closed on the left and open on the right, i.e. `f(0) = 1`, `f(1 – ε) = 1`, `f(1) = 2`, etc.

The instantaneous transfer rate is the slope of this function, but that’s not a useful thing to look at because the derivative is zero everywhere and it’s not differentiable at the discontinuities. (Which just means that if you measure over a small enough time window, then when there is no send in that window the rate is 0, and when there is a send in that window, you can make the rate arbitrarily high by making the window arbitrarily small.)

So what we want to look at instead is the average transfer rate over some time window, `f(t1) – f(t0) / (t1 - t0)`.

In the limit where `t1 - t0 → ∞`, this rate goes to 1 (for the function I plotted, but again the y-axis units are arbitrary). But rate limits in networks don’t work over infinite time windows so this is _also_ not so useful to look at.

So then what time window do we take? Well, if we take a time window of `t0=0, t1=1`, then we get `f(1) – f(0) / (1 – 0) = (2 – 1) / 1 = 1`. So we only count one rising edge inside the window. But if we take the limit `ε → 0` for the window `t0=–ε, t1=1`, then we get `f(–ε) = 0` so the rate is 2. So for a window of size 1, the rate is 1, but if we make the window the tiniest amount larger so it can fit two rising edges, the rate is suddenly 2.

If we plot the average transfer rate over a window `t0=0, t1=x`, it looks like this:

transfer-rate

> starting at 2x and gradually trending downwards to the threshold

So sure, it looks like the rate is high and then drops down to the target rate, but that’s an artifact caused by how we define the rate, the pattern of how we transfer does not change over time! If we only evaluate this rate function at integer times then we find a rate of 1 everywhere, but if we evaluate it at an ε past integers then we find a rate of `1 + 1/x` and we have this “gradually trending downwards towards the threshold”.

Now for my next magic trick, let’s look at the average transfer rate over a window `ε → 0 t0=ε, t1=x`, so now this initial rising edge is not part of the window. Then we get this:

rate-window-epsilon

Again, when we evaluate it at integer times then we find a rate of 1 everywhere (`x=0` does not count, the function is not defined for x < ε). But if we evaluate it at an ε before integers then we find a rate of `1 - 1/x`. Now it starts below the threshold and it’s gradually trending up!

So when you say

> but doing that you can send 2x the bucket size over the first transfer window.

I think we agree on what the send pattern should be, it’s just that we are talking about different windows to measure the rate over. If the bucket starts out empty, then initially we have to wait and cumulative bytes sent looks like this:

cumulative-sent

Which is exactly the same graph as before, just shifted in time, so the rate should also be the same, just shifted in time.

Comment (DavidVentura) [reply src/ratelimiter.rs]: Plotting the instant transfer rate (rate over a very small window) along with the average helped me understand that there are only two rates:
- While bucket is full, we transfer at the max bandwidth of the underlying connection
- While the bucket is "empty" (filling at its standard rate) we transfer at the max bandwidth that the ratelimiter allows

When the max bandwidth is over the target bandwidth (3.5mbps, 3mbps respectively here):
image




Starting with an empty bucket for comparison
image

shows the same behavior, only that we are indeed blocked until the bucket refills enough for the first transfer

Comment (ruuda) [reply src/ratelimiter.rs]: I think I assumed that the capacity of the bucket is on the same order of magnitude as what we send in one send call. If it’s smaller then we wouldn’t send anything, if it’s much larger then yes initially we are not limiting and sending as fast as the network allows. That’s a good way to determine the capacity of the bucket then, it should be about one chunk.

File: src/main.rs
Comment (ruuda) [inline src/main.rs:319]: Is the underscore to avoid shadowing the the `rl` Arc/Mutex? A prefix underscore in Rust usually means that a variable is intentionally not used, but that is not the case here, I would rename the outer `rl` to `rl_mutex` or something. Or even `limiter_mutex`, why leave the reader guessing at what RL means if we can just spell it out?

File: src/main.rs
Comment (ruuda) [inline src/main.rs:358]: Now we always count a full chunk, but at the end of the file we send less. When there are few big files this doesn’t matter, but when there are many small files we might limit more than intended. Probably for that case you don’t need Fastsync though so I think it’s fine for now, but also this is the kind of thing that is hard to find later on. Let’s at least add a `FIXME` comment, or alternatively we could pass an `Option<&mut RateLimiter>` to `send_one` and make it consume as much as it sent (or put the amount in `SendResult` and consume it here).

File: src/ratelimiter.rs
Comment (ruuda) [inline src/ratelimiter.rs:26]: This implicitly sets the bucket capacity at 1 second × bytes-per-second. If the chunk size is larger than one second of data, it means there will never be enough `bytes_available()` to send. (At least not in the current approach where we don’t adjust the chunk size.) Maybe it doesn’t happen in this case because the minimum is 1 MB/s and the chunk size is smaller than that, but then somebody later changes the chunk size or switches the limiter unit to kB/s and then it subtly breaks. I would add an explicit `capacity_bytes` field and initialize it to `MAX_CHUNK_LEN`.

File: src/ratelimiter.rs
Comment (ruuda) [inline src/ratelimiter.rs:34]: I would either put in an assert or use `saturating_sub` (depending on whether we want to crash on or silence a bug). Now if we make a programming error and consume more than what was available, in debug mode it will crash but in release it will overflow.

**Comment by yannickhilber** [comment]
Worked for me, max bandwith was enforced (100 mbps + 8 threads):
```
root@solana12:/opt/validators/solana/ledger# RUST_BACKTRACE=1 /root/fastsync send 100.87.67.117:4440 --max-bandwidth-mbps 100 snapshot-290328809-7xfSXni2M15KwUeZBZtDPuKt7M498DK5PZx4ZMQj7hFR.tar.zst
---
root@solana16:/opt/validators/solana/ledger# /root/fastsync recv 100.87.67.117:4440 8
```

https://monitor.prod.chorus1.net/d/rYdddlPWk/node-exporter?orgId=1&var-DS_PROMETHEUS=default&var-job=node_exporter&var-node=100.72.229.111:9100&var-diskdevices=%5Ba-z%5D%2B%7Cnvme%5B0-9%5D%2Bn%5B0-9%5D%2B&from=1726578000000&to=1726582500000

image

**Comment by yannickhilber** [comment]
Worked for me, max bandwith was enforced (100 mbps + 8 threads).

File: src/main.rs
Comment (DavidVentura) [reply src/main.rs:319]: `Option::insert` returns the current value in the option, which I don't need -- and is always `None`
I've renamed the `rl` variables, it's much clearer now

File: src/ratelimiter.rs
Comment (DavidVentura) [reply src/ratelimiter.rs]: much better now, thanks!

File: src/main.rs
Comment (DavidVentura) [reply src/main.rs]: This will mean that we must keep the ratelimiter lock while sending, I've implemented it at b99024e3c4e7bd80b841d5fe29dcde76c7ad05c5, let me know what you think

File: src/ratelimiter.rs
Comment (DavidVentura) [reply src/ratelimiter.rs:34]: asserted!

File: src/ratelimiter.rs
Comment (ruuda) [inline src/ratelimiter.rs:136]: We could make `time_until_bytes_available` return an `Option<Duration>`, where `None` means “you requested more than the bucket’s capacity”, but then at the call site it would be a programming error anyway, hmm. We could put in an assert that the request is not larger than the capacity, probably it’s probably better to crash the program than to wait indefinitely.

**Review by ruuda — APPROVED** [review]
LGTM, nice!

<sub>
I think with waiting under the lock, we might have a fairness problem if only one thread sends, but actually we already have this fairness problem, nothing guarantees that all threads send at the same rate, but so far it doesn’t seem to be a problem so let’s ignore this for now.
</sub>

**Pull request merged**
